# Analyze Electric Vehicle Stations in Python

- [View Solution Notebook](./solution.html)
- [View Project Page](https://www.codecademy.com/projects/practice/analyze-electric-vehicle-stations-with-python)

## Task Group 1 - Import and Explore

### Task 1

Import the CSV file `stations.csv` and assign it to the variable `stations`.

In [7]:
import pandas as pd
stations = pd.read_csv('stations.csv')
# show output
stations.head()


fuel state       owner   access  number_of_stations
0  biodiesel    AL  government  private                   8
1  biodiesel    AL     private   public                   2
2  biodiesel    AR  government  private                   1
3  biodiesel    AR     private   public                  16
4  biodiesel    AZ  government  private                  74

<details>
    <summary style="display:list-item; font-size:16px; color:blue;"><i>What is the structure of this dataset? Toggle to check!</i></summary>

Each row of the dataset counts the number of stations corresponding to certain categories of fuel, ownership, location, and access.
    
For example, let's look at the first row. The first row counts the number of
    
    - biodiesel stations
    - in Alabama
    - owned by the government
    - with restricted/private access

The number in the last column tells us there are 8 stations matching these properties!

</details>

### Task 2

When we start working with a new dataset, it's a good idea to get some summaries of the different columns, so that we know what kinds of values they contain.

Call `.value_counts()` on `fuel` to see the different kinds of fuel included in the dataset.

In [2]:
stations['fuel'].value_counts()

electric                  260
compressed natural gas    210
propane                   105
ethanol                    97
biodiesel                  70
liquefied natural gas      50
hydrogen                   27
Name: fuel, dtype: int64

### Task 3

Call `.value_counts()` on `owner` to see the different kinds of owners included in the dataset.

In [3]:
stations['owner'].value_counts()

private            413
government         288
utility company    108
joint               10
Name: owner, dtype: int64

### Task 4

Call `.value_counts()` on `access` to see the different kinds of access included in the dataset.

In [4]:
stations['access'].value_counts()

public     415
private    404
Name: access, dtype: int64

<details>
    <summary style="display:list-item; font-size:16px; color:blue;"><i>What did we discover using value counts? Toggle to check!</i></summary>

Using `.value_counts()` tells us the different types of stations in the dataset. For example, we now know that there are both public-access and private-access stations in the dataset.
    
It is important to remember that `.value_counts()` only tells us how many *rows* in the dataset contain a certain value. For example, from Task 4 we know that there are 415 rows in the dataset that contain `public` in the `access` column. This does *not* correspond to 415 stations, since each row can have a large `number_of_stations`.

</details>

### Task 5

Call `.describe()` on `number_of_stations`.

In [5]:
stations['number_of_stations'].describe()

count     819.000000
mean       34.741148
std       118.665095
min         1.000000
25%         2.000000
50%         5.000000
75%        20.500000
max      2423.000000
Name: number_of_stations, dtype: float64

### Task 6

There's a pretty large maximum in the output to Task 5. Sort `stations` by `number_of_stations` from largest to smallest. What do the top 5 rows have in common?

In [8]:
stations.sort_values(by='number_of_stations', ascending=False)

fuel state            owner   access  \
301                electric    CA          private   public   
456                electric    NY          private   public   
500                electric    TX          private   public   
329                electric    FL          private   public   
510                electric    VA          private   public   
..                      ...   ...              ...      ...   
623                 ethanol    TN          private  private   
285                electric    AL       government   public   
83   compressed natural gas    AZ       government   public   
627                 ethanol    UT          private   public   
117  compressed natural gas    GA  utility company   public   

     number_of_stations  
301                2423  
456                1173  
500                 778  
329                 659  
510                 497  
..                  ...  
623                   1  
285                   1  
83                    1  
627                   1  
117                   1  

[819 rows x 5 columns]

### Task 7

Sort `stations` by `number_of_stations` from smallest to largest. What do the top 5 rows of the new sorted DataFrame have in common?

In [9]:
stations.sort_values(by='number_of_stations', ascending=True)

fuel state       owner   access  number_of_stations
194  compressed natural gas    NJ       joint   public                   1
203  compressed natural gas    NV  government  private                   1
432                electric    NH  government  private                   1
208  compressed natural gas    NY       joint   public                   1
730                 propane    CT     private  private                   1
..                      ...   ...         ...      ...                 ...
510                electric    VA     private   public                 497
329                electric    FL     private   public                 659
500                electric    TX     private   public                 778
456                electric    NY     private   public                1173
301                electric    CA     private   public                2423

[819 rows x 5 columns]

<details>
    <summary style="display:list-item; font-size:16px; color:blue;"><i>What did we discover using sorting? Toggle to check!</i></summary>

The five largest rows are all public-access, privately-owned, and electric. 

We can't draw too many conclusions from the five smallest rows, because there may be more than just these five rows with 1 station.

</details>

## Task Group 2 - Public Access Electric Charging

## Task 8

Electric vehicles have become more and more crucial to plans around climate change. Let's take a closer look at stations that anyone can access.

Create a Boolean mask that is `True` for any row of `stations` where `access` is public.

In [12]:
is_public_access = stations['access'] == 'public'

### Task 9

Create a Boolean mask that is `True` for any row of `stations` where `fuel` is electric.

In [13]:
is_electric_fuel = stations['fuel'] == 'electric'

### Task 10

Use the Boolean masks from Tasks 8 and 9 to filter `stations` down to only rows that are both public-access and electric-fuel. Assign the result to the variable `public_electric`.

In [16]:
public_electric = stations[is_electric_fuel & is_public_access]

### Task 11

Sort `public_electric` by `number_of_stations` from smallest to largest. Display the top 5 rows (corresponding to the smallest numbers of stations.)

In [19]:
public_electric.sort_values(by='number_of_stations').head(5)

fuel state            owner  access  number_of_stations
454  electric    NY            joint  public                   1
473  electric    OR  utility company  public                   1
317  electric    DC       government  public                   1
458  electric    NY  utility company  public                   1
479  electric    RI       government  public                   1

### Task 12

While we can't be sure that this trend continues, it certainly looks as if publicly-owned (government/utility) stations are less common than privately owned (which we saw in Task 6).

Let's compare privately- and publicly-owned stations. Create a Boolean mask that is `True` for each row of `public_electric` where `owner` is private.

In [36]:
is_private_owner = public_electric['owner'] == 'private'

### Task 13

Use the Boolean mask from Task 12 to filter `public_electric` down to only privately-owned rows. Assign the result to the variable `privately_owned`.

In [37]:
privately_owned = public_electric[is_private_owner]

### Task 14

Let's check how many states have privately-owned, publicly-accessible electric charging stations. Call `.describe()` on the `state` column of `privately_owned`.

In [38]:
privately_owned['state'].describe()

count     51
unique    51
top       AK
freq       1
Name: state, dtype: object

<details>
    <summary style="display:list-item; font-size:16px; color:blue;"><i>What did we discover in Task 14? Toggle to check!</i></summary>

There are 51 unique values in the `privately_owned` state column. So it looks like all 50 states and (likely) the District of Columbia have at least one privately-owned, publicly-accessible electric charging station. 
    
We can't be 100% sure without actually looking at the data: it is possible that one state abbreviation got misspelled, producing a count of 51 distinct entries without actually having 51 distinct entries. But this is good initial evidence!
</details>

### Task 15

Let's compare this to the publicly-owned stations. Use the Boolean mask you created in Task 12 to filter `public_electric` down to only rows with *non*-private ownership. Assign the result to the variable `not_privately_owned`.

In [40]:
not_privately_owned = public_electric[~ is_private_owner]

### Task 16

Let's check how many states have publicly-owned, publicly-accessible electric charging stations. Call `.describe()` on the `state` column of `not_privately_owned`.

In [44]:
not_privately_owned['state'].describe()
not_privately_owned

fuel state            owner  access  number_of_stations
281  electric    AK       government  public                   6
283  electric    AK  utility company  public                   3
285  electric    AL       government  public                   1
290  electric    AR       government  public                   8
293  electric    AR  utility company  public                   1
..        ...   ...              ...     ...                 ...
525  electric    WA  utility company  public                   4
527  electric    WI       government  public                   5
531  electric    WI  utility company  public                   1
533  electric    WV       government  public                  11
537  electric    WY       government  public                   8

[83 rows x 5 columns]

<details>
    <summary style="display:list-item; font-size:16px; color:blue;"><i>What did we discover in Task 16? Toggle to check!</i></summary>

There are only 49 unique entries, so it is likely that there are two states (possibly one state and the District of Columbia) that have *zero* publicly-owned and publicly-accessible electric charging stations.

</details>

### Task 17

Let's investigate publicly-owned stations a bit further. Call `.describe()` on the `number_of_stations` column of `not_privately_owned`.

In [45]:
not_privately_owned['number_of_stations'].describe()

count     83.000000
mean      17.927711
std       44.294409
min        1.000000
25%        1.000000
50%        6.000000
75%       17.000000
max      361.000000
Name: number_of_stations, dtype: float64

### Task 18

The maximum number of stations is quite a bit larger than the 75th percentile. Let's look at the rows between these values. 

Create a Boolean mask that is true in each row of `not_privately_owned` where the `number_of_stations` is bigger than 17 (the 75th percentile).

In [50]:
not_privately_owned[not_privately_owned['number_of_stations'] > 17].sort_values(by='number_of_stations')

fuel state            owner  access  number_of_stations
423  electric    NC  utility company  public                  18
521  electric    WA       government  public                  19
398  electric    MI  utility company  public                  19
508  electric    VA       government  public                  19
367  electric    KS  utility company  public                  21
378  electric    MA       government  public                  22
394  electric    MI       government  public                  22
438  electric    NJ       government  public                  26
384  electric    MD       government  public                  30
390  electric    ME       government  public                  31
331  electric    FL  utility company  public                  31
327  electric    FL       government  public                  32
460  electric    OH       government  public                  39
303  electric    CA  utility company  public                  40
475  electric    PA       government  public                  46
418  electric    NC       government  public                  50
305  electric    CO       government  public                  59
484  electric    SC       government  public                  63
311  electric    CT       government  public                 106
453  electric    NY       government  public                 145
299  electric    CA       government  public                 361

### Task 19

Filter `not_privately_owned` down to only rows with `number_of_stations` bigger than 17. Assign the result to the variable `above_17`.

In [51]:
above_17 = not_privately_owned[not_privately_owned['number_of_stations'] > 17].sort_values(by='number_of_stations')

### Task 20

Sort `above_17` by `number_of_stations` and output the entirety of the result.

In [52]:
above_17

fuel state            owner  access  number_of_stations
423  electric    NC  utility company  public                  18
521  electric    WA       government  public                  19
398  electric    MI  utility company  public                  19
508  electric    VA       government  public                  19
367  electric    KS  utility company  public                  21
378  electric    MA       government  public                  22
394  electric    MI       government  public                  22
438  electric    NJ       government  public                  26
384  electric    MD       government  public                  30
390  electric    ME       government  public                  31
331  electric    FL  utility company  public                  31
327  electric    FL       government  public                  32
460  electric    OH       government  public                  39
303  electric    CA  utility company  public                  40
475  electric    PA       government  public                  46
418  electric    NC       government  public                  50
305  electric    CO       government  public                  59
484  electric    SC       government  public                  63
311  electric    CT       government  public                 106
453  electric    NY       government  public                 145
299  electric    CA       government  public                 361

<details>
    <summary style="display:list-item; font-size:16px; color:blue;"><i>What did we discover in Task 20? Toggle to check!</i></summary>

There's no immediately obvious geographic pattern, though it is interesting that California has both the most government-owned and the most utility-owned stations. However, we'd have to compare against population to see if that's just a factor of California's size.
    
Combining datasets (like `stations` with a population dataset) is beyond the scope of this project -- but is a powerful feature of pandas that you'll learn if you keep going!

</details>

## Task Group 3 - West Coast

### Task 21

In Task 20 we found that `CA` has 
1. the largest number of government-owned public-access electric stations
2. the largest number of utility-owned public-access electric stations

In Task 6, the full sorted DataFrame shows that `CA` also has the most privately-owned public-access electric stations. 

What about the rest of the west coast?

Create two Boolean masks:
- the first should be `True` whenever the `state` column of `public_electric` is `WA` (Washington)
- the second should be `True` whenever the `state` column of `public_electric` is `OR` (Oregon)

In [53]:
is_in_WA = public_electric['state'] == 'WA'
is_in_OR = public_electric['state'] == 'OR'

### Task 22

Filter `public_electric` down to only those rows where `state` is either `WA` or `OR`. Assign the result to the variable `WA_or_OR`.

In [56]:
WA_or_OR = public_electric[is_in_WA | is_in_OR]

### Task 23

Sort `WA_or_OR` first by `owner` and then, within each `owner`, by `number_of_stations`. Display the full output.

In [57]:
WA_or_OR.sort_values('owner').sort_values('number_of_stations')

fuel state            owner  access  number_of_stations
473  electric    OR  utility company  public                   1
525  electric    WA  utility company  public                   4
469  electric    OR       government  public                  16
521  electric    WA       government  public                  19
471  electric    OR          private  public                 282
523  electric    WA          private  public                 402

<details>
    <summary style="display:list-item; font-size:16px; color:blue;"><i>What did we discover in Task 23? Toggle to check!</i></summary>

Both WA and OR have significantly more privately-owned stations than publicly-owned (among publicly-accessible stations.) At the time this project was created, the population of WA was about double the population of OR (slightly below, in fact). Do you think the proportions of stations make sense?

</details>

That's the end of the project, but you can always add more cells below to explore the dataset further!